src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -0,0 +1,103 @@
+        private readonly IActionInvokerFactory _invokerFactory;
+        private readonly IActionContextAccessor _actionContextAccessor;
+        private readonly List<Endpoint> _endpoints;
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            IActionInvokerFactory invokerFactory)
+            : this(actions, invokerFactory, actionContextAccessor: null)
+        {
+        }
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            IActionInvokerFactory invokerFactory,
+            IActionContextAccessor actionContextAccessor)
+        {
+            if (actions == null)
+            {
+                throw new ArgumentNullException(nameof(actions));
+            }
+
+            if (invokerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(invokerFactory));
+            }
+
+            _actions = actions;
+            _invokerFactory = invokerFactory;
+
+            // The IActionContextAccessor is optional. We want to avoid the overhead of using CallContext
+            // if possible.
+            _actionContextAccessor = actionContextAccessor;
+
+            _endpoints = new List<Endpoint>();
+
+            InitializeEndpoints();
+        }
+
+        private void InitializeEndpoints()
+        {
+            // note: this code has haxxx. This will only work in some constrained scenarios
+            foreach (var action in _actions.ActionDescriptors.Items)
+            {
+                if (action.AttributeRouteInfo == null)
+                {
+                    // Action does not have an attribute route
+                    continue;
+                }
+
+                RequestDelegate invokerDelegate = (context) =>
+                {
+                    var values = context.Features.Get<IEndpointFeature>().Values;
+                    var routeData = new RouteData();
+                    foreach (var kvp in values)
+                    {
+                        routeData.Values.Add(kvp.Key, kvp.Value);
+                    }
+
+                    var actionContext = new ActionContext(context, routeData, action);
+                    if (_actionContextAccessor != null)
+                    {
+                        _actionContextAccessor.ActionContext = actionContext;
+                    }
+
+                    var invoker = _invokerFactory.CreateInvoker(actionContext);
+                    return invoker.InvokeAsync();
+                };
+
+                var descriptors = action.FilterDescriptors.OrderBy(f => f, FilterDescriptorOrderComparer.Comparer).Select(f => f.Filter).ToArray();
+                var metadataCollection = new EndpointMetadataCollection(descriptors);
+
+                _endpoints.Add(new MatcherEndpoint(
+                    next => invokerDelegate,
+                    "/" + action.AttributeRouteInfo.Template,
```

Comment:
I found I needed to prefix attribute templates with `/` before they matched. Something I'm missing here?

Reply:
hmm, dunno - might be a bug in the matcher that it doesn't sanitize these

Reply:
Is there a good way to debug routing source from within a MVC test?

Reply:
newp

Reply:
- Modify MVC repo's `build\dependencies.props` and find the Routing related package version and replace them with version that you find here `<repo-base-path>\routing\artifacts\build`. The version of locally built version would be `2.2.0-preview1-t000`.
- Also update MVC repo's `build\sources.props` to have a new RestoreSource pointing to your local routing artifacts folder `<repo-base-path>\routing\artifacts\build`